# Quality control for whole-genome sequencing data

This repository compiles a series of scripts to perform thorough quality control (QC) on whole-genome sequencing data in a VCF format. 
This pipeline builds on the initial QC methods used in the paper:

Acosta-Uribe, J., Aguill√≥n, D., Cochran, J.N. et al. A neurodegenerative disease landscape of rare mutations in Colombia due to founder effects. Genome Med 14, 27 (2022). https://doi.org/10.1186/s13073-022-01035-9

The scripts presented here are provided as a QC pipeline guide, rather than a plug and play script. There are steps that require a 'manual evaluation', and as many scripts, it needs to be tuned according to the user needs.


## Description

An in-depth paragraph about your project and overview of use.

## Getting Started

### Dependencies

* bcftools
* vcftools
* plink


### Installing

* How/where to download your program
* Any modifications needed to be made to files/folders

### Executing program

* How to run the program
* Step-by-step bullets
```
code blocks for commands
```

## Help

Any advise for common problems or issues.
```
command to run if program contains helper info
```

## Authors

Contributors names and contact info

ex. Dominique Pizzie  
ex. [@DomPizzie](https://twitter.com/dompizzie)

## Version History

* 0.2
    * Various bug fixes and optimizations
    * See [commit change]() or See [release history]()
* 0.1
    * Initial Release

## License

This project is licensed under the [NAME HERE] License - see the LICENSE.md file for details

## Acknowledgments

Inspiration, code snippets, etc.
* [awesome-readme](https://github.com/matiassingers/awesome-readme)
* [PurpleBooth](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2)
* [dbader](https://github.com/dbader/readme-template)
* [zenorocha](https://gist.github.com/zenorocha/4526327)
* [fvcproductions](https://gist.github.com/fvcproductions/1bfc2d4aecb01a834b46)
